File: 2023/go/pkg/day3/problem2.go
package day3

import (
	"fmt"
	"strconv"
	"strings"
	"unicode"
)

type gear struct {
	x, y int
}

func SolveProblem2(lines []string) string {
	sum := 0
	gears := map[gear]([]int){}

	for i, line := range lines {
		var temp strings.Builder

		for j, r := range line {
			if unicode.IsDigit(r) {
				temp.WriteRune(r)
			}

			if (!unicode.IsDigit(r) || j == len(line)-1) && temp.Len() > 0 {
				number, err := strconv.Atoi(temp.String())

				if err != nil {
					panic(err)
				}

				adjacent, gear := isAdjacentToGear(j-temp.Len(), i, temp.String(), lines)

				if adjacent {
					gears[gear] = append(gears[gear], number)
				}

				temp.Reset()
			}
		}
	}

	for _, gear := range gears {
		if len(gear) == 2 {
			sum += gear[0] * gear[1]
		}
	}

	return fmt.Sprint(sum)
}

func isAdjacentToGear(x, y int, num string, lines []string) (bool, gear) {
	for dy := -1; dy <= 1; dy++ {
		for dx := -1; dx <= len(num); dx++ {
			ny, nx := y+dy, x+dx
			if ny >= 0 && ny < len(lines) &&
				nx >= 0 && nx < len(lines[ny]) &&
				lines[ny][nx] == '*' {
				return true, gear{nx, ny}
			}
		}
	}
	return false, gear{}
}

File: 2023/rust/src/day3/problem2.rs
use std::collections::HashMap;

#[derive(Hash, Eq, PartialEq)]
struct Gear {
    x: usize,
    y: usize,
}

pub fn solve(lines: Vec<&str>) -> String {
    let mut sum = 0;
    let mut gears: HashMap<Gear, Vec<i32>> = HashMap::new();

    for (i, line) in lines.iter().enumerate() {
        let mut temp = String::new();

        for (j, char) in line.chars().enumerate() {
            if char.is_digit(10) {
                temp.push(char);
            }

            if (!char.is_digit(10) || j == line.len() - 1) && temp.len() > 0 {
                if let Some(gear) = is_adjecent_to_gear(&temp, j - temp.len(), i, &lines) {
                    let num = temp.parse::<i32>().expect("Failed to parse number :(");

                    match gears.get_mut(&gear) {
                        Some(nums) => nums.push(num),
                        None => {
                            gears.insert(gear, vec![num]);
                        }
                    }
                }

                temp.clear();
            }
        }
    }

    for (_, nums) in gears {
        if nums.len() == 2 {
            sum += nums[0] * nums[1];
        }
    }

    return sum.to_string();
}

fn is_adjecent_to_gear(num: &String, x: usize, y: usize, lines: &Vec<&str>) -> Option<Gear> {
    for dy in -1..2 {
        for dx in -1..num.len() as i32 + 1 {
            let x = x as i32 + dx;
            let y = y as i32 + dy;

            if y < 0 || x < 0 {
                continue;
            }

            let x = x as usize;
            let y = y as usize;

            if y >= lines.len() || x >= lines[y].len() {
                continue;
            }

            let char = lines[y].chars().nth(x).unwrap();

            if char == '*' {
                return Some(Gear { x, y });
            }
        }
    }

    None
}
